File: monitor/monitor.go
// This program demonstrates the use of "monitor" (mutex + condition varialbles) in the classic producer-consumer problem.

package main

import (
	"fmt"
	"sync"
)

type RingBuffer struct {
	buffer     []interface{}
	readP      int
	writeP     int
	mutex      sync.Mutex
	dataAvail  *sync.Cond
	spaceAvail *sync.Cond
}

func NewRingBuffer(size int) *RingBuffer {
	rb := RingBuffer{buffer: make([]interface{}, size+1)}
	rb.dataAvail = sync.NewCond(&rb.mutex)
	rb.spaceAvail = sync.NewCond(&rb.mutex)
	return &rb
}

func (rb *RingBuffer) Put(dataUnit interface{}) {
	func() {
		rb.mutex.Lock()
		defer rb.mutex.Unlock()

		for (rb.writeP+1)%len(rb.buffer) == rb.readP { // while full
			rb.spaceAvail.Wait()
		}

		rb.buffer[rb.writeP] = dataUnit
		rb.writeP = (rb.writeP + 1) % len(rb.buffer)
	}()
	rb.dataAvail.Signal()
}

func (rb *RingBuffer) Get() interface{} {
	dataUnit := func() interface{} {
		rb.mutex.Lock()
		defer rb.mutex.Unlock()

		for rb.readP == rb.writeP { // while empty
			rb.dataAvail.Wait()
		}

		dataUnit := rb.buffer[rb.readP]
		rb.buffer[rb.readP] = nil // In case the value held is a pointer, tell the garbage collector it is no longer needed.
		rb.readP = (rb.readP + 1) % len(rb.buffer)

		return dataUnit
	}()
	rb.spaceAvail.Signal()
	return dataUnit
}

type theDataUnit struct {
	sourceThreadId int
	data           rune
}

func producer(rb *RingBuffer, threadId int, wg *sync.WaitGroup) {
	for i := 0; i < 10; i++ {
		dataUnit := &theDataUnit{threadId, 'a' + int32(i)}
		rb.Put(dataUnit)
	}
	wg.Done()
}

func consumer(rb *RingBuffer, threadId int, wg *sync.WaitGroup) {
	for i := 0; i < 8; i++ {
		dataUnit := rb.Get().(*theDataUnit)
		fmt.Printf("Thread %d got %c from thread %d\n", threadId, dataUnit.data, dataUnit.sourceThreadId)
	}
	wg.Done()
}

const numProducers = 4
const numConsumers = 5

func main() {
	rb := NewRingBuffer(3)

	var wg sync.WaitGroup
	wg.Add(numProducers)
	wg.Add(numConsumers)

	for i := 0; i < numProducers; i++ {
		go producer(rb, i, &wg)
	}
	for i := 0; i < numConsumers; i++ {
		go consumer(rb, i+numProducers, &wg)
	}

	wg.Wait()
}

File: monitor/monitor.java
// This program demonstrates the use of a "monitor" (a mutex + condition varialbles) in the classic producer-consumer problem.

import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.locks.Condition;

public class monitor {
	private static class RingBuffer<T> {
		private final T[] buffer;
		private int readP;
		private int writeP;

		private final Lock lock = new ReentrantLock();
		private final Condition dataAvail = lock.newCondition();
		private final Condition spaceAvail = lock.newCondition();

		public RingBuffer(int size) {
			@SuppressWarnings("unchecked")
			T[] buffer = (T[]) new Object[size + 1];
			// We cannot create generic arrays in Java. So we create an Object[] and pretend it to be a T[].
			// The cast is OK because T is actually Object at run-time, due to type erasure of Java's generics.
			// We can only use annotations on declarations. So we use a temporary variable here.

			this.buffer = buffer;
		}

		public void put(T dataUnit) {
			lock.lock();
			try {
				while ((writeP + 1) % buffer.length == readP)
					spaceAvail.awaitUninterruptibly();

				buffer[writeP] = dataUnit;
				writeP = (writeP + 1) % buffer.length;

				dataAvail.signal(); // Java implementations usually require the lock is locked when signaling.
			}
			finally {
				lock.unlock();
			}
		}

		public T get() {
			lock.lock();
			try {
				while (readP == writeP)
					dataAvail.awaitUninterruptibly();

				T dataUnit = buffer[readP];
				buffer[readP] = null; // Tell the garbage collector this reference is no longer needed.
				readP = (readP + 1) % buffer.length;

				spaceAvail.signal(); // Java implementations usually require the lock is locked when signaling.

				return dataUnit;
			}
			finally {
				lock.unlock();
			}
		}
	}

	private static class DataUnit {
		public final int sourceThreadId;
		public final char data;
		public DataUnit(int sourceThreadId, char data) {
			this.sourceThreadId = sourceThreadId;
			this.data = data;
		}
	}

	private static class Producer implements Runnable {
		private final RingBuffer<DataUnit> rb;
		private final int threadId;
		public Producer(RingBuffer<DataUnit> rb, int threadId) {
			this.rb = rb;
			this.threadId = threadId;
		}
		@Override
		public void run() {
			for (int i=0; i<10; ++i) {
				DataUnit dataUnit = new DataUnit(threadId, (char)((int)'a' + i));
				rb.put(dataUnit);
			}
		}
	}

	private static class Consumer implements Runnable {
		private final RingBuffer<DataUnit> rb;
		private final int threadId;
		public Consumer(RingBuffer<DataUnit> rb, int threadId) {
			this.rb = rb;
			this.threadId = threadId;
		}
		@Override
		public void run() {
			for (int i=0; i<8; ++i) {
				DataUnit dataUnit = rb.get();
				String output = String.format("Thread %d got %c from thread %d", threadId, dataUnit.data, dataUnit.sourceThreadId);
				System.out.println(output);
			}
		}
	}

	private static final int NUM_PRODUCERS = 4;
	private static final int NUM_CONSUMERS = 5;

	public static void main(String[] args) {
		RingBuffer<DataUnit> rb = new RingBuffer<DataUnit>(3);

		Thread[] producers = new Thread[NUM_PRODUCERS];
		for (int i=0; i<NUM_PRODUCERS; ++i) {
			producers[i] = new Thread(new Producer(rb, i));
			producers[i].start();
		}
		Thread[] consumers = new Thread[NUM_CONSUMERS];
		for (int i=0; i<NUM_CONSUMERS; ++i) {
			consumers[i] = new Thread(new Consumer(rb, i + NUM_PRODUCERS));
			consumers[i].start();
		}

		// There is no need to join the threads.
	}
}
